### PR TITLE
Internal Plugins

### PR DESCRIPTION
**Pull request by cbednarski**
This is a port of a similar feature in packer: https://github.com/mitchellh/packer/pull/2608

Instead of compiling a separate binary for each provider / provisioner, all of the plugins are now compiled into the main `terraform` binary and invoked via the `internal-plugin` command. This is invisible to the user and happens automatically.

The change to the plugin system is designed to search for traditional, separate plugin binaries first, which means you can still build custom plugins or even override the internal plugin behavior.

There is also a code generator to automatically compile in new plugins as they are added. I project, though I have not tested, that it also should be faster to dev/test by commenting out parts of `command/internal_plugin_list.go` that you aren't working on. When I was developing this branch I used a subset of plugins and my `go build` times were around 23 seconds.

### Why do this?

There is a lot of code duplicated between the individual plugin binaries. Compiling a single binary which includes everything saves space and takes less time to compile. Here are some numbers:

#### On the current 0.7 dev branch

```
(dev-0.7) 0$ time make dev
...
real	3m14.365s
user	20m28.031s
sys	1m57.843s

(dev-0.7) 0$ du -hs bin/
751M	bin/
```

#### On the new internal-plugins branch

```
(f-internal-plugins) 0$ time make dev
...
real	0m59.019s
user	3m32.701s
sys	0m18.518s

(f-internal-plugins) 0$ du -hs bin/
130M	bin/
```

#### Summary

In summary there is approximately 70% reduction in time to run `make dev` and approximately 80% reduction in the size of the binary/ies. These gains should also carry into the cross-compilation / release process.

I also compared times for `make`, which runs tests. There is no noticeable difference in time here, though there is a slight difference in behavior. Please try out your normal dev/test workflows and let me know if you run into any slowdowns.

Also since this is mostly toolchain changes I'm not sure of the appropriate way to test this; I'm open to suggestions for this.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -106,6 +108,27 @@ func (c *Config) Discover() error {
 		}
 	}
 
+	// Finally, we'll fill in any internal plugins that haven't already been
+	// discovered on disk.
```
Comment (phinze) [inline config.go]: The upgrade process here will be a little weird, since we'll be distributing a single binary to users who have all the plugins on their path.

Do you have any experience from this with Packer? We should figure out what the expected upgrade UX is in this case.

File: config.go [hunk 2]
```diff
@@ -106,6 +108,27 @@ func (c *Config) Discover() error {
 		}
```
Comment (phinze) [inline config.go]: In the above code - I believe the lookup order is actually backwards from what we want - PWD should override builtin. Had a dive into this semi-recently with a user who wanted to build a branch with newer AWS resources and use it on atlas.

Not saying you have to fix that in the same go - just mentioning here that we should get this fixed too mostly for my own remembering. :grinning:

File: config.go [hunk 3]
```diff
@@ -285,6 +308,12 @@ func pluginCmd(path string) *exec.Cmd {
...
 	}
 
+	// No plugin binary found, so try to use an internal plugin.
+	if strings.Contains(path, command.TFSPACE) {
+		parts := strings.Split(path, command.TFSPACE)
+		return exec.Command(parts[0], parts[1:]...)
+	}
```
Comment (phinze) [inline config.go]: Does this show up in the debug output? I'd like to make sure w/ `TF_LOG=trace` you can see where each plugin came from.

File: scripts/build.sh [hunk 4]
```diff
@@ -41,16 +41,8 @@ gox \
...
-    -output "pkg/{{.OS}}_{{.Arch}}/terraform-{{.Dir}}" \
-    $(go list ./... | grep -v /vendor/)
-
-# Make sure "terraform-terraform" is renamed properly
-for PLATFORM in $(find ./pkg -mindepth 1 -maxdepth 1 -type d); do
-    set +e
-    mv ${PLATFORM}/terraform-terraform.exe ${PLATFORM}/terraform.exe 2>/dev/null
-    mv ${PLATFORM}/terraform-terraform ${PLATFORM}/terraform 2>/dev/null
-    set -e
-done
```
Comment (phinze) [inline scripts/build.sh]: OMG this is exciting!!!!!!

File: config.go (hunk 1 above)
Comment (cbednarski) [inline config.go]: Yeah this is a good point. There was a bit of pain around this in Packer so we added a note to the [download page](https://www.packer.io/downloads.html) and also called this out in the changelog.

Ideally we could detect that the plugin is old but there is currently no plugin versioning so this isn't something we can check.

What I could do is add versioning going forward and we can complain in the UI if the plugin doesn't have a version specified. This might involve a change to hashicorp/go-plugin.

File: scripts/generate-plugins.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	path := "./builtin/providers"
+	typeName := "Provider"
+	typeID := "terraform.ResourceProvider"
+	return discoverTypesInPath(path, typeID, typeName)
+}
+
+func discoverProvisioners() ([]plugin, error) {
+	path := "./builtin/provisioners"
+	typeID := "ResourceProvisioner"
+	typeName := ""
```
Comment (phinze) [inline scripts/generate-plugins.go]: Can you make the local vars in these two funcs the same order?

File: scripts/generate-plugins.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+func deriveName(root, full string) string {
+	short, _ := filepath.Rel(root, full)
+	bits := strings.Split(short, string(os.PathSeparator))
+	return strings.Join(bits, "-")
+}
+
+// deriveImport will build a unique import identifier based on packageName and
+// the result of deriveName()
+//
+// This will be something like    -> awsprovider
```
Comment (phinze) [inline scripts/generate-plugins.go]: Is there supposed to be something before the arrow here?

File: config.go (hunk 2 above)
Comment (cbednarski) [inline config.go]: I think there are three levels of precedent (I ran into this in packer, also). From most to least preferred:

1. Same directory as the binary (`/usr/local/bin`)
2. Config dir (`~/.terraform/plugins`) ??
3. Builtin

I don't know of anyone who actually uses 2.

The order in the code is 2, 3, 1. It's slightly confusing but I feel doing the "fallback" approach is easier than changing the earlier code. We can change this if you prefer the priority / order in the code are the same but I think it _is_ actually working the way we want now.

File: scripts/generate-plugins.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+						})
+						// The AST stops parsing when we return false. Once we
+						// find the symbol we want we can stop parsing.
+						return false
+					}
+				}
+				return true
+			})
+		}
+	}
```
Comment (phinze) [inline scripts/generate-plugins.go]: We might want to add a unit test or three around this behavior at some point, but since we'll be exercising this code quite a bit I'm cool going with it for now.

**Comment by phinze** [comment]
Great great great work on this @cbednarski! Just made a pass with a bunch of inline little bitty bits, but this is going to be aaaaaa mazing!

File: config.go (hunk 2 above)
Comment (cbednarski) [inline config.go]: Oops, I see what you mean with respect to PWD.

File: config.go (hunk 1 above)
Comment (phinze) [inline config.go]: Ah yes that reminds me we're actually bumping the plugin API version in 0.7, so old plugins will definitely blow up w/ a version mismatch. We can play around with the code path and layer on some nicer messages explaining what to do. :+1:

File: scripts/generate-plugins.go (hunk 6 above)
Comment (cbednarski) [inline scripts/generate-plugins.go]: Yeah some of these are confusing thanks to some combination of copy+paste+find+replace. I will make a second pass here on comments.

File: config.go (hunk 1 above)
Comment (cbednarski) [inline config.go]: Can you point to where the plugin API version is specified? I can probably put a version check in that part of the code. I don't think packer has that feature.

Comment (phinze) [inline config.go]: https://github.com/hashicorp/terraform/blob/2df53f283dbcb5b4278705ef85e1987d7c2ecb11/plugin/server.go#L18-L21

Comment (phinze) [inline config.go]: Hm maybe that's not it. I'll swing back around to this.

File: config.go (hunk 3 above)
Comment (cbednarski) [inline config.go]: For internal plugins this shows up in the debug log:

```
go-dynect 2016/04/11 16:33:25 [DEBUG] plugin: starting plugin: /Users/cbednarski/go/bin/terraform []string{"/Users/cbednarski/go/bin/terraform", "internal-plugin", "provider", "aws"}
go-dynect 2016/04/11 16:33:25 [DEBUG] plugin: starting plugin: /Users/cbednarski/go/bin/terraform []string{"/Users/cbednarski/go/bin/terraform", "internal-plugin", "provisioner", "remote-exec"}
```

Comment (phinze) [inline config.go]: Cool that works - just as long as we can see it I'm happy!

Comment (cbednarski) [inline config.go]: Just took a quick look -- this is easy to do for internal plugins but slightly more complicated for disk-based plugins because of the override behavior during discovery. I can still add this, though.

File: scripts/generate-plugins.go (hunk 7 above)
Comment (cbednarski) [inline scripts/generate-plugins.go]: Yeah I think there are some places I can add tests here. IMO negative cases are complicated for this. Also the generated code gets checked in so we should be able to see if anything aberrant happens.

File: scripts/generate-plugins.go (hunk 5 above)
Comment (cbednarski) [inline scripts/generate-plugins.go]: This is updated but the comment diff may not reflect the change.

File: config.go (hunk 2 above)
Comment (cbednarski) [inline config.go]: This is done.

File: config.go (hunk 1 above)
Comment (cbednarski) [inline config.go]: I took a brief look at go-plugin but it doesn't seem easy to add version in a backwards-compatible way so I'm going to defer this. It may be possible to do something like pass an ENV alongside the magic cookie but I think that leaves enforcement of the version constraint up to the plugin instead of the main binary, which seems problematic.

File: scripts/generate-plugins.go (hunk 7 above)
Comment (cbednarski) [inline scripts/generate-plugins.go]: Tests have been added.

File: config.go (hunk 3 above)
Comment (cbednarski) [inline config.go]: This involves a slightly more complicated refactor so I'm inclined to defer this. There is currently logging when internal plugins are used but to differentiate between the various types of disk-based plugins will require some changes to the discovery logic.

**Comment by cbednarski** [comment]
@phinze I've updated the PR based on your feedback. I've added tests. They are not exhaustive, but should verify the AST parsing code as well as some of the command passing logic. Let me know if you have any outstanding concerns!
